src/rules/core/no-id-references-for-responses-rule.js

```diff
@@ -0,0 +1,74 @@
+const Rule = require('../rule');
+const ValidationErrorCategory = require('../../errors/validation-error-category');
+const ValidationErrorSeverity = require('../../errors/validation-error-severity');
+const validationErrorType = require('../../errors/validation-error-type');
+
+/** @typedef {import('../../classes/model-node').ModelNodeType} ModelNode */
+
+class NoIdReferencesForResponsesRule extends Rule {
+  constructor(options) {
+    super(options);
+    this.targetValidationModes = [
+      'C1Response',
+      'C2Response',
+      'PResponse',
+      'BResponse',
+      'OrdersFeed',
+      'OrderStatus',
+    ];
+    this.targetFields = { OrderItem: ['acceptedOffer', 'orderedItem'] };
+    this.meta = {
+      name: 'NoIdReferencesForResponsesRule',
+      description: 'Validates that acceptedOffer and orderedItem are not ID references and not objects for responses (C1, C2 etc)',
+      tests: {
+        default: {
+          description: `Raises a failure if the acceptedOffer or orderedItem within the OrderItem of a response is a URL 
+          (ie a reference to the object and not the object itself)`,
+          message: 'For responses, {{field}} must not be an compact ID reference, but the object representing the data itself',
+          sampleValues: {
+            field: 'acceptedOffer',
+          },
+          category: ValidationErrorCategory.CONFORMANCE,
+          severity: ValidationErrorSeverity.FAILURE,
+          type: validationErrorType.FIELD_SHOUlD_NOT_BE_ID_REFERENCE,
+        },
+      },
+    };
+  }
+
+  /**
+   *
+   * @param {ModelNode} node
+   * @param {string} field
+   */
+  validateField(node, field) {
+    // Don't do this check for models that we don't actually have a spec for or for models that aren't JSON-LD
+    if (!node.model.hasSpecification || !node.model.isJsonLd) {
+      return [];
+    }
+
+    const errors = [];
+    const fieldValue = node.getValue(field);
+
+    if (typeof fieldValue !== 'object') {
```

Comment:
This will pass for arrays (e.g. `typeof ["https://an-id-reference.com/123"]` is `object`) and `null`.

You could use the lodash `isPlainObject(..)` function instead.

---

Another thing is: Do you also need to check if it's the right kind of object or will another validator rule do that for you?

For example, with this project compiled, if you run `validate(..)` in C1Response mode where `orderedItem[0].orderedItem = {}`, will Validator correctly mark this as invalid?

Reply:
Have changed the first part!

The second question, there is a `NoEmptyValuesRule` that marks any empty objects are invalid, so I think we are covered!